### DISCUSSION TITLE
OptionsProvider with constructor

### DISCUSSION THREAD

**Discussion by langert** [pimcore/pimcore#17717] [Q&A / Support] [answered]
Hi,
I use a custom OptionsProvider (implements SelectOptionsProviderInterface) for providing options for a select field for pimcore Dataobjects. 

In the pimcore backend the keys for the options are not translated. Now I tried to use TranslatorInterfacein my OptionsProvider to translate the keys. But while pimcore is loading the Optionsprovider in Backend when calling a Dataobject, there occures an error "too few arguments".

I configured the @Symfony\Contracts\Translation\TranslatorInterface as argument in my service.yaml but still same error.

Is it bad practice to have a constructor in a OptionsProvider ?

**Comment by langert** [comment]
It seems that pimcore diretly instatiates the provider class without using dependency injection.

`Message: Too few arguments to function MyBundle\CoreBundle\OptionsProvider\AbstractSelectPropertyProvider::__construct(), 0 passed in /var/www/html/www/vendor/pimcore/pimcore/models/DataObject/ClassDefinition/Helper/ClassResolver.php on line 35 and exactly 1 expected`

How can I access the ParameterBagInterface in an OptionProvider if I can not use dependency injection?

**Reply by jdreesen** [reply]
Please show your config/code.

**Comment by langert** [comment]
I have an Abstract Class
```
class AbstractSelectPropertyProvider implements SelectOptionsProviderInterface {

    static $defaultValue = null;
    static $hasStaticOptions = false;

    public function __construct(protected ?ParameterBagInterface $params) {
    }

    public function getOptions($context, $fieldDefinition): array {
    ...
   }
}
```
and some classes that extend it

```
class CharacterProvider extends AbstractSelectPropertyProvider {

}
```
in service.yaml I have

```
MyBundle\CoreBundle\OptionsProvider\AbstractSelectPropertyProvider:
    arguments: ["@parameter_bag"]
```

I have a configured a classificationStore with a keyDefinition "character" where I set the CharacterProvider as the Options Provider Class.

Opening a DataObject in pimcore backend haveing the classification store, I get the error.

**Reply by jdreesen** [reply]
Do you also have a service definition for the `CharacterProvider`?

I think you have to specify the argument there (or extend the `AbstractSelectPropertyProvider` service via `parent`, see: https://symfony.com/doc/current/service_container/parent_services.html)

**Comment by langert** [comment]
same problem if I have a service definition like this - that looks unneccessary duplicated to me...

```
  MyBundle\CoreBundle\OptionsProvider\AbstractSelectPropertyProvider:
    abstract: true
    arguments: ["@parameter_bag"]
    public: true

  MyBundle\CoreBundle\OptionsProvider\:
    resource: "../src/OptionsProvider"
    parent: MyBundle\CoreBundle\OptionsProvider\AbstractSelectPropertyProvider
    arguments: ["@parameter_bag"]
    public: true
```

**Reply by jdreesen** [reply, marked as answer]
And the service name of the options provider in your classification store starts withan `@`? (it must be the service ID starting with an `@`).

**Comment by langert** [comment]
jackpot! I used the path by namespace and not the service 
changed from \MyBundle\.... to @MyBundle\... solved the issue.
Thanks for guiding me to that point!
